staging/src/k8s.io/client-go/util/workqueue/delaying_queue.go

```diff
@@ -264,19 +450,216 @@ func (q *delayingType) waitingLoop() {
+func (q *delayingType) calculateNextReadyAt(w *waitForPriorityQueue) {
+	if w.Len() == 0 {
+		// remove timer when the queue becomes empty
+		if q.nextReadyHead != nil {
+			q.nextReadyAtTimerCh = make(<-chan time.Time)
+			if q.nextReadyAtTimer != nil {
+				q.nextReadyAtTimer.Stop()
+				q.nextReadyAtTimer = nil
+			}
+			q.nextReadyHead = nil
+		}
+		return
+	}
+
+	// Keep the existing nextReadyAt timer if the head duration has not changed.
+	entry := w.Peek().(*waitFor)
+	if q.nextReadyHead != nil && q.nextReadyHead.readyAt == entry.readyAt {
+		return
+	}
+
+	// Recalculate the nextReadyAtTimer
+	q.nextReadyAtTimerCh = make(<-chan time.Time)
+	if q.nextReadyAtTimer != nil {
+		q.nextReadyAtTimer.Stop()
+		q.nextReadyAtTimer = nil
+	}
+
+	now := q.clock.Now()
+	q.nextReadyAtTimer = q.clock.NewTimer(entry.readyAt.Sub(now))
+	q.nextReadyAtTimerCh = q.nextReadyAtTimer.C()
+	q.nextReadyHead = entry
+	if q.logger != nil {
+		q.log(fmt.Sprintf("waitingLoop: calculated NEXT-READY-AT timer " + entry.readyAt.Sub(now).String()))
+	}
+}
+
+// closeAllWaitForFills loops through all the channels in waitForQueueFilled and closes them before resetting the list.
+func (q *delayingType) closeAllWaitForFills() {
+	if len(q.waitForQueueFilled) > 0 {
+		if q.logger != nil {
+			q.log(fmt.Sprintf("waitingLoop: closing %d waitForQueueFilled channels", len(q.waitForQueueFilled)))
+		}
+		for _, ch := range q.waitForQueueFilled {
+			close(ch)
+		}
+		q.waitForQueueFilled = make([]chan int, 0)
+	}
+}
+
+// processArrivingMessage decides whether to accept or drop an entry based on its waitFor settings and state of the queue.
+// It implements the following behaviours:
+//   - An entry is dropped if it is already actively queued (much like Add() on the Interface)
+//   - Delayed items are inserted into the 'waiting' queue according to their Waiting option (TakeShorter by default)
+//   - Immediate items will be dropped when the TakeLonger Waiting option is set and they are already present in 'waiting'.
+//   - Immediate items will pre-empt/remove items in the 'waiting' queue unless the PermitActiveAndWaiting option is set.
+//
+// processArrivingMessage will return entries
+func (q *delayingType) processArrivingMessage(w *waitForPriorityQueue, knownEntries map[t]*waitFor, entry *waitFor) {
+	// handle the different message actions
+	switch entry.action {
+	case waitForActionExists:
+		if entry.returnCh == nil {
+			return
+		}
+		r := 0
+		_, exists := knownEntries[entry.data]
+		if exists {
+			r = 1
+		}
+		if q.logger != nil {
+			q.log(fmt.Sprintf("waitingLoop: return exists result %t for %v", exists, entry.data))
+		}
+		entry.returnCh <- r
+		close(entry.returnCh)
+		return
+	case waitForActionLength:
+		if entry.returnCh == nil {
+			return
+		}
+		r := len(knownEntries)
+		if q.logger != nil {
+			q.log("waitingLoop: waiting queue length: " + strconv.Itoa(r))
+		}
+		entry.returnCh <- r
+		close(entry.returnCh)
+		return
+	case waitForActionQueueFilled:
```

Comment:
I am still doubts if we should have so much testing logic inside the queue, but it seems pretty harmless and without a significant overhead at this time.

Do you think it would help if we gated all the testing code with a gate? Could be a simple boolean debug/testing flag.

Also, changing `if q.logger != nil` to `if q.loggingEnabled` could also make the code a bit more readable. We could initialize the variable  according to the presence of the logger, but maybe we can do this under one flag (eg `q.testingEnabled`) since logging is used only in testing.

Reply:
Hmm, I have been wondering whether it would be better to expose these methods/actions to all clients instead - why shouldn't a client want to test for an item in the queue or request the queue length?  I was trying to keep the changes to the DelayingInterface minimal but as I said, some of these actions feel like primitives that clients might want to use.

Reply:
What do you think?  I'm exposing the new Exists method on Interface?

Reply:
I don't know, I would be careful about adding additional functionality to the API. If needed, we can try in a followup, where we can consider also things like the use cases and implementation